pure-js/mutual-tls-example/routes/index.js

```diff
@@ -0,0 +1,109 @@
+const JOSEWrapper = require('@idpartner/jose-wrapper');
+const { v4: uuidv4 } = require('uuid');
+const config = require('../config.json');
+
+const scope = 'openid offline_access email profile birthdate address';
+
+const state = crypto.randomBytes(16).toString('hex');
+const nonce = crypto.randomBytes(16).toString('hex');
+const verifier = base64.encode(crypto.randomBytes(32));
+const challenge = base64.encode(crypto.createHash('sha256').update(verifier).digest());
+
+// Load certificate and key for mutual TLS
+const clientCert = fs.readFileSync(`./certs/${config.client_id}.pem`);
+const clientKey = fs.readFileSync(`./certs/${config.client_id}.key`);
+
+const router = express.Router();
+
+router.get('/', async (_req, res, next) => {
+  try {
+    return res.render('index', { title: 'RP Mutual TLS Example using node-oidc-client', config });
+  } catch (error) {
+    return next(error);
+  }
+});
+
+router.get('/button/oauth', async (req, res, next) => {
+  try {
+    const { iss, idp_id: idpId, visitor_id: visitorId } = req.query;
+    if (iss) {
+      req.session.issuer = iss;
+      // Build query parameters for the authorization request
+      const queryParams = querystring.stringify({
+        client_id: config.client_id,
+        code_challenge: challenge,
+        code_challenge_method: 'S256',
+        identity_provider_id: idpId,
+        nonce: nonce,
+        prompt: 'consent',
+        redirect_uri: config.redirect_uri,
+        response_mode: 'jwt',
+        response_type: 'code',
+        scope: scope,
+        state: state,
+        'x-fapi-interaction-id': uuidv4()
+      });
+
+      // Redirect the user to the authorization URL
+      return res.redirect(`${iss}/auth?${queryParams}`);
+    } else {
+      // bank selection
+      return res.redirect(`https://auth-api.idpartner.com/oidc-proxy/auth/select-accounts?client_id=${config.client_id}&visitor_id=${visitorId}&scope=${scope}`)
+    }
+  } catch (error) {
+    return next(error);
+  }
+});
+
+router.get('/button/oauth/callback', async (req, res, next) => {
+  try {
+    // Fetch the OpenID Connect Discovery document from the provided issuer URL
+    const discoveryResponse = await axios.get(`${req.session.issuer}/.well-known/openid-configuration`);
+    const discoveryData = discoveryResponse.data;
+
+    // Obtain the endpoints from the discovery data
+    const mTlsUserinfoEndpoint = discoveryData.mtls_endpoint_aliases.userinfo_endpoint;
+    const mTlsTokenEndpoint = discoveryData.mtls_endpoint_aliases.token_endpoint;
+
+    const decodedToken = await JOSEWrapper.verifyJWS({ jws: req.query.response, issuerURL: req.session.issuer });
+
+    const payload = {
+      client_id: config.client_id,
+      code: decodedToken.code,
+      code_verifier: verifier,
+      grant_type: 'authorization_code',
+      redirect_uri: config.redirect_uri,
+    };
+
+    const httpsAgent = new https.Agent({
+      cert: clientCert,
+      key: clientKey,
+      rejectUnauthorized: true,
```

Comment:
This should be false

Reply:
we need to reject self-signed or invalid SSL certificates. this is more secure in true, any reason to change it?

Reply:
Sorry, I got confused. You are right. Please dismiss this comment. Thansk!

Reply:
discussed offline: Chsico asked me to remove this line